akka-projection-cassandra/src/test/scala/akka/projection/cassandra/CassandraContainerSpec.scala

```diff
@@ -0,0 +1,61 @@
+package akka.projection.cassandra
+
+import java.net.InetSocketAddress
+
+import akka.Done
+import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
+import akka.event.Logging
+import akka.stream.alpakka.cassandra.{ CassandraSessionSettings, CqlSessionProvider }
+import akka.stream.alpakka.cassandra.scaladsl.CassandraSession
+import com.datastax.oss.driver.api.core.CqlSession
+import com.datastax.oss.driver.internal.core.metadata.DefaultEndPoint
+import com.dimafeng.testcontainers.{ CassandraContainer, ForAllTestContainer }
+import org.scalatest.BeforeAndAfterAll
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+trait CassandraContainerSpec extends ForAllTestContainer with BeforeAndAfterAll { self: ScalaTestWithActorTestKit =>
+  override val container: CassandraContainer = CassandraContainer()
+
+  private val configPath = "akka.projection.cassandra"
+  protected var session: CassandraSession = _
+
+  override protected def beforeAll(): Unit = {
+    super.beforeAll()
+    session = cassandraSession
+  }
+
+  private def sessionProvider: CqlSessionProvider = {
+    val cqlSession: CqlSession = CqlSession.builder
+      .addContactEndPoint(
+        new DefaultEndPoint(
+          InetSocketAddress
+            .createUnresolved(
+              container.cassandraContainer.getContainerIpAddress,
+              container.cassandraContainer.getFirstMappedPort.intValue())))
+      .withLocalDatacenter("datacenter1")
+      .build()
+
+    new CqlSessionProvider {
+      override def connect()(implicit ec: ExecutionContext): Future[CqlSession] = Future.successful(cqlSession)
+    }
+  }
+
+  // testcontainers facilitates the creation of a session. we should have an easier way to programmatically provide
+  // a test session in Alpakka Cassandra
+  private def cassandraSession: CassandraSession = {
+    val settings = CassandraSessionSettings(configPath)
+    val classic = system.classicSystem
+    val log = Logging(classic, classOf[CassandraSession])
+    val sessionDispatcher = system.settings.config.getConfig(settings.configPath).getString("session-dispatcher")
+    val executionContext = classic.dispatchers.lookup(sessionDispatcher)
+    new CassandraSession(
+      classic,
+      sessionProvider,
+      executionContext,
+      log,
+      metricsCategory = settings.metricsCategory,
+      init = settings.init.getOrElse(_ => Future.successful(Done)),
+      onClose = () => ())
+  }
```

Comment:
I know this whole approach is wrong, but I implemented this just to get things working.  I think Alpakka Cassandra needs more flexible options to override configuration at runtime.  Initially I tried providing an overridden typesafe `Config` object at runtime, but the `CassandraSessionRegistry` APIs only expect paths to config files, or a config object that contains a `session-provider` FQCN, but that's not flexible enough to override with runtime docker container ip and dynamic port information that testcontainers provides.  Maybe I'm missing something?

Reply:
As a workaround (until Alpakka Cassandra does not have this flexibility) we can use static values in `Config` and with Cassandra container run a `TcpProxy`. This approach described in Sergei's article https://bsideup.github.io/posts/debugging_containers/. 
I used this workaround in [draft PR](https://github.com/lagom/lagom/pull/2057) for migrating Lagom to testcontainers. https://github.com/lagom/lagom/blob/d011ba84bcc10d20c41cb224a54a995c5bd7c896/dev/cassandra-server/src/main/scala/com/lightbend/lagom/internal/cassandra/CassandraLauncher.scala#L52-L55

Reply:
As another idea to use a temporary file.
1. Start Cassandra container
2. Create a temporary config (from template) file with container params (IP, port)
3. Load `Config` from temporary file with fallback `Config` from classpath if needed.
4. Initialize a Cassandra session by `Config` from point 3.

Reply:
Then it's more complicated than thought. Maybe we should stick with the Docker Cassandra for now. Doesn't feel like the most important thing to use TestContainers if we already have something that is working.

Reply:
Also we can try use a hack for set static port to Cassandra container:
```scala
cassandraServer.getPortBindings.add(s"""${port}:${CassandraContainer.CQL_PORT}/${InternetProtocol.TCP.toDockerNotation}""")
```

Reply:
@ihostage Thanks for the suggestions.  I figured out that I just need to call container `start()` method myself in a new Alpakka Cassandra session provider.  `start()` will block until the container's internal readiness checks are green, and then I can generate new sessions.